Gradually decay assigned misbehavior over time

ptschip opened the pull request:
Decay misbehavior over a 4 hour window of time.

Also, updated a few other variables and made them atomic since we don't lock anything when we assign misbehavior.

ptschip committed:
Make nMisbehavior an atomic double

ptschip committed:
Decay nMisbehavior over a four hour window

ptschip committed:
Make fShouldBan atomic

ptschip committed:
Make nBanType atomic

ptschip committed:
Decay the misbehavior using the QT polltimer

This is so that in the QT peers tab the misbehavior will show up
correctly as decaying over time.

However, the user may not be running QT, but it doesn't matter
because the misbehavior is always decayed just before adding any
new misbehavior points and also before and ban would be enacted
based on the new points that were added.

ptschip committed:
Split out the updating of the misbeavior's to a different function

This clarifies the code.

ptschip committed:
Add tests for decaying misbehavior

ptschip commented:
Taking out of WIP:

I've tested this and verified that the misbehavior dynamically updates in QT, but if anyone wants to test this you can go into net.h and set the nMisbehavior{80}; so that when each peer connects it automatically has some misbehavior assinged.  Then let it run for an hour or so and check your peers misbehavior which should be less than when you started.

ptschip commented:
all tests passed although travis does not show that

sickpig commented:
Even thou the travis checks are failed, they passed, check

https://travis-ci.org/github/BitcoinUnlimited/BitcoinUnlimited/builds/669815884

gandrewstone commented on src/dosman.cpp:
```diff
@@ -273,18 +273,49 @@ void CDoSManager::Misbehaving(CNode *pNode, int howmuch, BanReason reason)
 
-    if (prior + howmuch >= nBanThreshold && prior < nBanThreshold)
+    // Add the new misbehavior and check whether to ban
+    double prior = pNode->nMisbehavior.load();
```
compare_exchange_weak allows the programmer to make arbitrary changes atomically.  But if you just want to atomically add, you can use the fetch_add or operator+= member functions.  This is a lot cleaner.
https://en.cppreference.com/w/cpp/atomic/atomic/operator_arith2
https://en.cppreference.com/w/cpp/atomic/atomic/fetch_add

ptschip replied:
fetch_add isn't available for floating point atomics until C++20

gandrewstone commented on src/qt/clientmodel.cpp:
```diff
@@ -118,6 +119,15 @@ void ClientModel::updateTimer2()
     graphenedata.FillGrapheneQuickStats(grapheneStats);
     Q_EMIT grapheneBlockPropagationStatsChanged(grapheneStats);
 
+    // Update all misbehaviors
```
I think you put this in the QT code only because the bitcoind code updates when misbehaving happens, and there is no other display of misbehaving.  However, we should instead add a misbehaving field to getpeerinfo (ideally also including the last error), since this will be very useful both in regression tests and in the live network. And so let's move this periodic update of misbehaviors to somewhere common.

ptschip replied:
I'll create a new thread for running these updates every 10 secs.  BTW, banscore is already in getpeerinfo.  As for showing the last banreason, that would be more difficult to do and I'd like to leave that for a different PR.

gandrewstone commented on src/dosman.cpp:
```diff
@@ -273,18 +273,49 @@ void CDoSManager::Misbehaving(CNode *pNode, int howmuch, BanReason reason)
+    while (true)
+    {
+        double nOldMisBehavior = pNode->nMisbehavior.load();
+        double nNewMisBehavior =
```
if (nOldMisBehavior == 0.0) break;

Doing this would skip the pow and the atomic ops for this common case... perhaps its an optimization without proof of bottleneck, but in this case its very simple to implement.

ptschip replied:
done

ptschip committed:
Break early if old misbehavior is already zero

ptschip committed:
Use an atomic load to assign the nMisbehaior for node stats

ptschip committed:
Create a new thread which updates misbehavior every 10 seconds

Remove the updating of misbehavior from the QT poll time and
run it in a seperate thread so that we always have correct and
updated misbehavior data which is also needed in getpeerinfo.

gandrewstone commented:
Note travis is green even though github did not update.
